torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -844,8 +858,8 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupMPI::recv(
   }
 
   return c10::make_intrusive<AsyncWork>(
```

Comment:
(This is prior to this PR and can be addressed in followup PRs) Would I be correct if I assume `AsyncWork` is only used in send/recv and allToAll, while all other API return `WorkMPI`? Why do we need two different work types?

Reply:
AFAIK `AsyncWork` is only used in send/recv, impls of Alltoall are using sync MPI ops and WorkMPI.

nonblocking MPI calls are used with Async work, and blocking with WorkMPI. I am not sure why we made a decision of using nonblocking versions of send/recv.

There is some guidance on how to use MPI send/recv:
https://www.mcs.anl.gov/research/projects/mpi/sendmode.html

Reply:
Do we plan to keep `WorkMPI` and `AsyncWork` separate or does it make sense to merge? IIUC, we plan to use thread-pools to mark async send/recv as ready. If that's the case, it will bring the use cases of AsyncWork closer to WorkMPI?

Reply:
correct. i think it'd make sense, in this case we won't have benefits of MPI_ISend/MPI_IRecv and should probably switch to their sync counterparts.